### PR TITLE
Bg/lwpmiopen 99 : CK's Conv + bias + activation kernel integration into MIOpen

### PR DESCRIPTION
**Pull request by bghimireamd**
* Integrate forward convolution bias activation fusion kernel from CK
* Update cba (convolution, bias, activation) test multiple layouts
* stride fix in TensorDescriptor for any layout

### CODE REVIEW

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]:
```suggestion
    if(problem.conv_problem.GetInDataType() != problem.conv_problem.GetWeightsDataType() ||
       problem.conv_problem.GetInDataType() != problem.conv_problem.GetOutDataType())
```

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]:
```suggestion
MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_CONV_CK_IGEMM_FWD_BIAS_ACTIV)
```
and use it in IsApplicable

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Is it possible provided that `problem = ctx.problem.GetConvProblem(0, conv::Direction::Forward)`?

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Is INT8 supported?

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: For unsupported types, this function cannot return anything meaningful. Shall we MIOPEN_THROW or assert at least?

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: index should be >= 0 as well

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: total_size is either initialized to "unknown" (-1) or dynamically computed in Init(). How it could be so that two instances have different total_size? Can you please provide an example scenario that leads to this mismatch?

Maybe we should throw in case of mismatch?

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (atamazov) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: [Important] Is it possible that `conv.size()` will change in the future as CK evolves (causing `add_device_conv2d_fwd_xdl_c_shuffle_bias_relu_nhwc_kyxc_nhwk_f16_instances()` to change its behavior)?

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Would `MIOPEN_THROW("Unsupported datatype")` be suffice?

**Comment by junliume** [comment]
@bghimireamd could you resolve discussions above? Ask @atamazov  for re-review since CI has passed now.

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: made the code change.

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Add check in IsApplicable for `MIOPEN_DEBUG_CONV_CK_IGEMM_FWD_BIAS_ACTIV`

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Agreed! so I removed that line.

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: not yet, just miopenHalf for now.

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: ended up just comparing kernel_id (name)

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: I believe so. As CK evolves `add_device_conv2d_fwd_xdl_c_shuffle_bias_relu_nhwc_kyxc_nhwk_f16_instances()` could add more kernels.

**Comment by bghimireamd** [comment]
@junliume Thank you. @atamazov made changes as per your comments.

File: Jenkinsfile
Comment (junliume) [inline Jenkinsfile]: @bghimireamd mmm ... I hope this is only a workaround and WIP, CMIIW but we cannot merge it this way as is.

Comment (bghimireamd) [reply Jenkinsfile]: @junliume Yes, this is a minor work around since it was causing my PR to[ stall](http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/bg%2FLWPMIOPEN-99/19/pipeline/17#step-257-log-2). I saw in [development branch](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/Jenkinsfile#L499) that it was commented. So, I did the same thing. It would be taken care during merging with updated [JD's branch](https://github.com/ROCmSoftwarePlatform/MIOpen/tree/jd/non_template_fusion_ops). 


So, once JD's branch is merged to development. Then I will pull the latest development. Once that is done then it would be ready to merge to development.

**Comment by JehandadKhan** [comment]
@bghimireamd Please follow up on this PR

File: driver/tensor_driver.hpp
Comment (JehandadKhan) [inline driver/tensor_driver.hpp]: Why is this required?

File: src/CMakeLists.txt
Comment (JehandadKhan) [inline src/CMakeLists.txt]:
```suggestion
    solver/conv_ck_igemm_fwd_bias_activ_fused.cpp
```

Comment (JehandadKhan) [reply src/CMakeLists.txt]: Please rename the file

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (JehandadKhan) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]:
```suggestion
 * Copyright (c) 2023 Advanced Micro Devices, Inc.
```

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (JehandadKhan) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: This does not align with the comment above? Perhaps you meant to use `GetTypeId` ?

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (JehandadKhan) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Please add an `MIOPEN_THROW` as the default case

File: test/gtest/cba.hpp
Comment (JehandadKhan) [inline test/gtest/cba.hpp]: Perhaps this should be fixed in the tensor class with a constructor that accepts layout as an argument ?

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (JehandadKhan) [inline src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Are there no instances for FP32 at all ?

**Review by JehandadKhan — DISMISSED** [review]
The new solver needs to be added to the static [fused solver list here ](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/aa087c18fdd154465a04a3f55062804fa97498bf/src/fusion.cpp#L395) so that the API can pick it up.

**Comment by JehandadKhan** [comment]
@iq136boy  Please help review

File: driver/tensor_driver.hpp
Comment (bghimireamd) [reply driver/tensor_driver.hpp]: I was using `tensor_driver.hpp` in our test and it includes `driver.hpp`. `driver.hpp` then includes `driver/random.hpp`. In `driver/randon.hpp` there is `GET_RAND` function. There is also another `GET_RAND` in `test/random.hpp`. This was causing multiple definition error. Two solutions : 

1) Avoid including `driver.hpp` that would avoid `driver/random.hpp` being included(current solution)
2) add `namespace test` in `test/random.hpp` around `GET_RAND` ( I like this one though )

File: src/CMakeLists.txt
Comment (bghimireamd) [reply src/CMakeLists.txt]: will do

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: will do

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: I will have to remove all the hacks with `unique_id` since CK will be fixing the `GetTyepString()`

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: will do

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (bghimireamd) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Currently CK is only supporting FP16 https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/library/src/tensor_operation_instance/gpu/conv2d_fwd_bias_relu/device_conv2d_fwd_xdl_c_shuffle_bias_relu_nhwc_kyxc_nhwk_f16_instance.cpp

File: test/gtest/cba.hpp
Comment (bghimireamd) [reply test/gtest/cba.hpp]: I agree! Let me fix this in constructor of [test/tensor_holder.hpp](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/test/tensor_holder.hpp)

File: driver/tensor_driver.hpp
Comment (JehandadKhan) [reply driver/tensor_driver.hpp]: I like 2 as well.

File: src/solver/conv_ck_igemm_fwd_bias_activ.cpp
Comment (JehandadKhan) [reply src/solver/conv_ck_igemm_fwd_bias_activ.cpp]: Lets not wait for that to happen and keep this PR functionally correct.

**Comment by bghimireamd** [comment]
Once PR #1957 is merged to develop all test should pass.

**Comment by atamazov** [comment]
@bghimireamd 
> Once PR #1957 is merged to develop all test should pass.

If this PR wants some newer CK, then it may worth changing git hash right here.

File: driver/driver.hpp
Comment (atamazov) [inline driver/driver.hpp]: [Q] Why cppcheck (or clang-tidy) wants this?

File: driver/random.hpp
Comment (atamazov) [inline driver/random.hpp]: [Q] Ditto

File: driver/random.hpp
Comment (atamazov) [inline driver/random.hpp]: [Q] Same

File: driver/tensor_driver.hpp
Comment (atamazov) [inline driver/tensor_driver.hpp]: [Q] Same

File: src/solver.cpp
Comment (atamazov) [inline src/solver.cpp]: 🔴 Please read comment at line 560!

File: src/CMakeLists.txt
Comment (atamazov) [inline src/CMakeLists.txt]: This list must be kept in sorted order.

File: test/gtest/cba.hpp
Comment (atamazov) [inline test/gtest/cba.hpp:85]: [Notice] 👌 Good catch. Is this fix important? If yes, then it might be worth creating a separate PR to make it distinct and speed up its consolidation.

File: test/tensor_holder.hpp
Comment (atamazov) [inline test/tensor_holder.hpp]: [Q] This looks like a patch to the ctor of miopen::TensorDescriptor. Maybe it would be better to fix it?

File: test/tensor_holder.hpp
Comment (atamazov) [inline test/tensor_holder.hpp]: [Q] Same

File: test/tensor_holder.hpp
Comment (atamazov) [inline test/tensor_holder.hpp]: [Q] Same

**Review by atamazov — CHANGES_REQUESTED** [review]
Partial review

File: test/random.hpp
Comment (atamazov) [inline test/random.hpp]: [Q] Do we really need this? GET_RAND is not included into the library and thus does not litter the global namespace.

File: test/gtest/cba.hpp
Comment (bghimireamd) [reply test/gtest/cba.hpp:85]: Since its just part of test/gtest code, creating a sperate PR might be overkill :smiley:

**Comment by junliume** [comment]
ping @JehandadKhan and @atamazov @iq136boy for review

File: src/solver/conv_ck_igemm_fwd_bias_activ_fused.cpp
Comment (JehandadKhan) [inline src/solver/conv_ck_igemm_fwd_bias_activ_fused.cpp:415]: Why is this throwing ? Shouldnt this just return false ?

**Review by JehandadKhan — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Pull request merged**
